src/lang.rs

```diff
@@ -424,9 +426,12 @@ impl std::ops::Add for Obj {
     type Output = EvalResult;
 
     fn add(self, rhs: Self) -> Self::Output {
-        match (self.as_double()?, rhs.as_double()?) {
+        if !(&self, &rhs).coercible_to_numeric() {
+            return Err(Signal::Error(Error::CannotBeCoercedToNumerics));
+        }
```

Comment:
What benefit does the `(Obj, Obj).coercible_to_numeric()` add here? To me it feels a bit un-rust-y. The `?` style coercion feels more natural to me.

Reply:
Yeah, I don't like it either. The problem is that the `Add` implementation for `Vector` (https://github.com/dgkf/R/blob/08e0c940c62e3d4ca226e74581473733f37efabc/src/object/vector/core.rs#L426-L438) does not return a `Result` object, which means that we cannot do `Ok(Obj::Vector((l + r)?))` here https://github.com/dgkf/R/blob/08e0c940c62e3d4ca226e74581473733f37efabc/src/lang.rs#L428. By checking for this convertibility before going into the `Add` implementation for `Vector` it ensures that the call does not panic, while still relying on the (I think relatively efficient) iterator-based conversion mechanism that you implemented here https://github.com/dgkf/R/blob/08e0c940c62e3d4ca226e74581473733f37efabc/src/object/vector/iterators.rs#L39-L60.

Ideally I think we should implement the conversion to the common minimally numeric type directly on `Obj`, but this would currently require an allocation of the vector, therefore disregarding the iterator-based conversion for `Vector`. If I were to add an `Iterator` variant to `RepType` in this PR, I think it would work.

Reply:
@dgkf do you think it makes sense to first implement `Iterator` for `RepType`?

Reply:
Actually, I think we don't really want the standard rust `Iterator` trait, but something where we can access element `i` before consuming elements `1, ..., i - 1`. Something like a `Vectorator` :smile:. I have implemented something like this in R for torch tensors, so I would feel comfortable giving this a try.